src/relentless/optimize/objective.py

```diff
@@ -353,15 +353,29 @@ def compute(self, variables, directory=None):
         # run simulation and use result to compute gradient
         try:
             sim = self.simulation.run(self.potentials, directory)
-            sim_ens = sim[self.thermo]["ensemble"]
+            if self._use_trajectory(self.target, self.thermo):
+                sim_ens = sim.directory.file(self.thermo.filename)
+            else:
+                if self.thermo.rdf is None:
+                    raise ValueError(
+                        "EnsembleAverage needs to compute RDF, specify parameters."
+                    )
+                sim_ens = sim[self.thermo]["ensemble"]
         finally:
             mpi.world.barrier()
             if tmp is not None:
                 tmp.cleanup()
 
         # compute gradient and result
         # relative entropy *value* is None
-        gradient = self.compute_gradient(sim_ens, variables)
+        gradient = None
+        if mpi.world.rank_is_root:
+            if self._use_trajectory(self.target, self.thermo):
+                gradient = self._compute_gradient_direct_average(sim_ens, variables)
+            else:
+                gradient = self.compute_gradient(sim_ens, variables)
+        mpi.world.barrier()
+        gradient = mpi.world.bcast(gradient)
```

Comment:
I think the problem is that you're broadcasting a dict that uses variable objects (not their names) as keys and so are actually different on different ranks. (A poor design choice we made long ago... #184).

It's a bit clunky, but what you can do is turn the `gradient` object into a list, broadcast the list, then reassign the list to the right keys.
```suggestion
        gradient_list = None
        if mpi.world.rank_is_root:
            if self._use_trajectory(self.target, self.thermo):
                gradient = self._compute_gradient_direct_average(sim_ens, variables)
            else:
                gradient = self.compute_gradient(sim_ens, variables)
            gradient_list = [gradient[var] for var in variables]
        gradient_list = mpi.world.bcast(gradient_list)
        gradient = {var: gradient_list[i] for i, var in enumerate(variables)}
```